clash-lib/src/Clash/Core/Evaluator/Models.hs

```diff
@@ -0,0 +1,305 @@
+import qualified Data.Map.Strict as Map
+import GHC.Generics (Generic)
+
+import Clash.Core.DataCon (DataCon)
+import Clash.Core.FreeVars (localFVsOfTerms, tyFVsOfTypes)
+import Clash.Core.Literal (Literal)
+import Clash.Core.Subst (extendTvSubstList, mkSubst, substTm)
+import Clash.Core.Term (Term(..), Pat, PrimInfo, TickInfo, mkApps)
+import Clash.Core.TyCon (TyConMap)
+import Clash.Core.Type (Type)
+import Clash.Core.Var (Id, IdScope(..), TyVar)
+import Clash.Core.VarEnv
+  ( InScopeSet, extendInScopeSet, mkInScopeSet
+  , VarEnv, delVarEnv, extendVarEnv, lookupVarEnv
+  , unionVarSet, uniqAway
+  )
+
+import Clash.Driver.Types (BindingMap, Binding(..))
+
+-- | An evaluator contains the basic functions that define partial evaluation.
+-- This consists of:
+--
+--   * evaluateWhnf, which evaluates terms to WHNF
+--   * quote, which recursively evaluates sub-terms and eta-expands to NF
+--
+-- Users of an evalautor can choose whether they convert terms to WHNF or NF,
+-- as the 'AsTerm' class (which converts back to Term) has instances for both.
+--
+data Evaluator = Evaluator
+  { evaluateWhnf :: Term -> State Env Value
+  , quote        :: Value -> State Env Nf
+  }
+
+-- | Evaluate a term to normal form without stopping at WHNF.
+--
+evaluateNf :: Evaluator -> Term -> State Env Nf
+evaluateNf (Evaluator e q) = e >=> q
+
+-- Local bindings are not stored in a VarEnv, as we still want to be
+-- able to access the keys (VarEnv simply indexes by unique).
+
+type EnvTyMap = Map TyVar Type
+type EnvTmMap = Map Id (Either Term Value)
+
+-- | EnvGlobals refers to global bindings in scope during evalaution, i.e.
+-- other top-level functions used in a program.
+--
+type EnvGlobals = VarEnv (Binding (Either Term Value))
+
+-- | EnvPrimsIO refers to the result of IO operations performed in primitives
+-- during evalaution e.g. creating a new ByteArray in the GHC frontend.
+--
+type EnvPrimsIO = (IntMap Value, Int)
+
+-- | An Environment contains all in scope terms and types while evaluating.
+-- This consists of
+--
+--   * local types and terms from the term being evaluated
+--   * top-level definitions in scope (i.e. global bindings)
+--   * IO results of primitive operations
+--   * in scope uniques and a means of generating uniques
+--
+-- When running the partial evaluator at different times, it may be desirable
+-- to keep the globals (some of which may have been partially evaluated) and
+-- the results of primitives in IO for future runs.
+--
+data Env = Env
+  { envTypes   :: !EnvTyMap
+  , envLocals  :: !EnvTmMap
+  , envGlobals :: !EnvGlobals
+  , envPrimsIO :: !EnvPrimsIO
+  , envTcMap   :: TyConMap
+  , envInScope :: InScopeSet
+  , envSupply  :: Supply
+  }
+
+instance Show Env where
+  show env = show (envTypes env, envLocals env, envPrimsIO env)
+
+-- This instance cannot be derived with Generic, as neither InScopeSet
+-- or Supply have instances for NFData. Instead, we use this instance
+-- which forces all fields which can be forced.
```

Comment:
`Supply` has no `NFData` instance as it's an infinitely big construct, but maybe add it for `InScopeSet`?

Reply:
As in add the instance to the type?